src/include/session.h

```diff
@@ -96,6 +96,12 @@ struct __wt_session_impl {
 	size_t	  scratch_cached;	/* Scratch bytes cached */
 #ifdef HAVE_DIAGNOSTIC
 	/*
+	 * Variables used to look for violations of the contract that a
+	 * session is only used by a single session at once.
+	 */
+	volatile size_t api_tid;
+	volatile size_t api_enter_refcnt;
```

Comment:
I don't think `volatile` is necessary.

If you want to be sure you catch threads racing, my guess is you want a publish when the thread-ID is set, does that sound correct to you?

Reply:
I added volatile because I didn't want the compiler to optimize out any code related to the variables. I think that's valid - I don't know specifically how the compiler could avoid re-reading a variable (which is the canonical example where volatile is useful), but my understanding is that it's best to mark any variable shared across threads as volatile.

I think a publish might be necessary - without it the macro might trigger a false positive, I didn't add it because I figured that any application that shares session structures amongst threads should have an explicit protocol in place to hand out sessions, so not protecting this assertion could help uncover application usage problems as well.